### ISSUE TITLE
Timeout not configurable for rancher2_bootstrap resource

### ISSUE THREAD

**Issue by wonderhoss** [rancher/terraform-provider-rancher2#477]
I am trying to provision a rancher instance on Azure compute and then using the `rancher2_bootstrap` resource to configure it.

This works as expected except during the initial terraform apply, the `rancher2_bootstrap` creation times out before the Azure instance is ready.
Waiting a couple more minutes and then re-applying terraform results in the boostrap being completed successfully.

It doesn't look like the timeout is configurable on this resource?

My module currently looks like this:

```
provider "rancher2" {
  alias = "bootstrap"

  api_url   = "https://${module.rancher_server.public_ip}"
  bootstrap = true
  insecure = true
}

resource "rancher2_bootstrap" "admin" {
  provider = rancher2.bootstrap
}
```

This results in:
```
[...]
rancher2_bootstrap.admin: Creating...
rancher2_bootstrap.admin: Still creating... [10s elapsed]
rancher2_bootstrap.admin: Still creating... [20s elapsed]
rancher2_bootstrap.admin: Still creating... [30s elapsed]
rancher2_bootstrap.admin: Still creating... [40s elapsed]
rancher2_bootstrap.admin: Still creating... [50s elapsed]
rancher2_bootstrap.admin: Still creating... [1m0s elapsed]
rancher2_bootstrap.admin: Still creating... [1m10s elapsed]
rancher2_bootstrap.admin: Still creating... [1m20s elapsed]
rancher2_bootstrap.admin: Still creating... [1m30s elapsed]
rancher2_bootstrap.admin: Still creating... [1m40s elapsed]
rancher2_bootstrap.admin: Still creating... [1m50s elapsed]
rancher2_bootstrap.admin: Still creating... [2m0s elapsed]
rancher2_bootstrap.admin: Still creating... [2m10s elapsed]
rancher2_bootstrap.admin: Still creating... [2m20s elapsed]
rancher2_bootstrap.admin: Still creating... [2m30s elapsed]
rancher2_bootstrap.admin: Still creating... [2m40s elapsed]

Error: Rancher is not ready: Doing get: Get https://<redacted>/ping: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
```

After a few minutes, another terraform apply yields:
```
[...]
rancher2_bootstrap.admin: Creating...
rancher2_bootstrap.admin: Still creating... [10s elapsed]
rancher2_bootstrap.admin: Creation complete after 13s [id=<redacted>]

Apply complete! Resources: 1 added, 0 changed, 0 destroyed.
```

Is there anything I can do with the latest version (I'm using `1.10.3`) to overcome this?

**Comment by rawmind0** [comment]
You could configure the number of retries of provider connection, https://registry.terraform.io/providers/rancher/rancher2/latest/docs#retries

Anyway, if you are using modules, the best way would be don't init the rancher2 provider until the required resource are up and running. For example, if other other modules are deploying vm's and k8s cluster, don't init rancher2 provider untile the k8s cluster is up and ready to connect. This shoudl be done managing tf modules vars/dependencies.

**Comment by wonderhoss** [comment]
Ah, I'm going to give `retries` a go. Thank you very much!

I'm not sure module dependencies would work here because the time is actually spent in node boot, after the terraform run has already successfully completed.
The node is configured through `user_data` and just takes some time to install the systemd unit and then pull docker images.
There's no provisioner that would track this setup.

Unless I'm missing something, the module spinning up the VM would still complete successfully and the dependant Rancher module would start before the node boot sequence is complete. There's nothing waiting for that to finish.

**Comment by wonderhoss** [comment]
Sadly `retries` doesn't quite solve the problem. It does allow the provider to continue trying while the node boots, but unfortunately it has revealed a nasty race condition in Rancher's docker setup itself.
It begins listening on its socket before it is ready to receive bootstrapping.
The error received depends on timing, but two examples I have encountered are:

```
Error: [ERROR] Connecting with user/pass: Bad response statusCode [403]. Status [403 Forbidden]. Body: [baseType=error, code=Forbidden, message=settings.management.cattle.io "k8s-version" is forbidden: User "user-4f4rl" cannot get resource "settings" in API group "management.cattle.io" at the cluster scope] from [https://51.141.102.99/v3/settings/k8s-version]
```

and

```
Error: [ERROR] Login with admin user: Doing  user logging: error Unauthorized
```


Again, wait a minute and run apply again and the bootstrap completes successfully.

Has anyone come across a solution for this? My aim is to provision and bootstrap a Rancher server within terraform without using any provisioners. That should be possible...?

**Comment by rawmind0** [comment]
@gargath , for sure it's possible, but as mentioned, you better sync your tf modules. The module that install Rancher helm chart,  shouldn't return until Rancher is up and ready. The own provider is deploying a Rancher HA on a k3s cluster and doing bootstrap on acceptance test suite, https://github.com/rancher/terraform-provider-rancher2/blob/master/scripts/start_rancher.sh#L103

**Comment by wonderhoss** [comment]
@rawmind0 I think that might be where the disconnect is.
I'm following a pattern I've used in the past where the application setup is done purely in user_data.
Terraform provisions a VM with a user_data script that sets up a systemd unit to run Rancher. There is no module doing helm stuff and there is no way (that I'm aware of) for terraform to wait for the instance to complete cloud-init.

Is there no way, then, to make this work?

Given that terraform is meant to be eventually consistent, I feel it would be much safer if Rancher wouldn't throw errors like that if bootstrapping is attempted before it is ready.

**Comment by rawmind0** [comment]
The provider is waiting to auth until `<RANCHER_URL>/ping` is answering `pong`. At this point, it's going to do login. If getting error on login then throw error logically. What do you propose, retrying login forever no matter if you got `Unauthorized`  error?? 

Again, by the moment this can work but you should sync. Terraform can wait for cloud-init to finish adding this at vm's definition
```
  provisioner "remote-exec" {
    inline = [
      "cloud-init status --wait"
    ]

    connection {
      type        = "ssh"
      host        = self.public_ip
      user        = var.node_username
      private_key = file(var.ssh_key_file)
    }
  }
```

**Comment by wonderhoss** [comment]
As I said, I'd like to avoid using a provisioner for this.

What I'm proposing, though, is not to retry forever.

Instead, `/ping` should only start returning success once the server is ready for bootstrapping.
The errors I've encountered clearly isn't a "real" unauthorised but rather simply a timing issue with the internals of Rancher getting all set up.
Wait a minute and the errors go away.

Currently I'm using Terraform's time wait resource to add a five minute delay. That works just fine, but is horrible. ><

**Comment by rawmind0** [comment]
> Instead, /ping should only start returning success once the server is ready for bootstrapping.

Agreed with you on that but this issue should be opened at rancher/rancher.

As workaround, i've added a fix on PR #487 , 3 retried if user/pass login fails when bootstrap.
